Modules/_testcapimodule.c

```diff
@@ -7622,13 +7655,47 @@ PyInit__testcapi(void)
     }
     PyModule_AddObject(m, "HeapCTypeMetaclass", HeapCTypeMetaclass);
 
+    PyObject *HeapCTypeMetaclassCustomNew = PyType_FromMetaModuleAndSpec(
+        &PyType_Type, m, &HeapCTypeMetaclassCustomNew_spec, (PyObject *) &PyType_Type);
+    if (HeapCTypeMetaclassCustomNew == NULL) {
+        return NULL;
+    }
+    PyModule_AddObject(m, "HeapCTypeMetaclassCustomNew", HeapCTypeMetaclassCustomNew);
+
     PyObject *HeapCTypeViaMetaclass = PyType_FromMetaModuleAndSpec(
         (PyTypeObject *) HeapCTypeMetaclass, m, &HeapCTypeViaMetaclass_spec, NULL);
     if (HeapCTypeViaMetaclass == NULL) {
         return NULL;
     }
     PyModule_AddObject(m, "HeapCTypeViaMetaclass", HeapCTypeViaMetaclass);
 
+    PyObject *HeapCTypeViaMetaclassCustomNew = PyType_FromMetaModuleAndSpec(
+        (PyTypeObject *) HeapCTypeMetaclassCustomNew, m, &HeapCTypeViaMetaclassCustomNew_spec, NULL);
+
+    int observed_exception = 0;
+    if (HeapCTypeViaMetaclassCustomNew == NULL && PyErr_ExceptionMatches(PyExc_SystemError)) {
+        /* This is expected to fail -- the metaclass overrides tp_new,
+           which is not supported by PyType_FromMetaModuleAndSpec */
+        PyObject *ptype, *pvalue, *ptraceback;
+        PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+        if (pvalue) {
+            PyObject *pstr = PyObject_Str(pvalue);
+            if (pstr) {
+                const char *msg = PyUnicode_AsUTF8(pstr);
+                observed_exception = strcmp(msg, "Metaclasses with custom tp_new are not supported.") == 0;
+            }
+            Py_XDECREF(pstr);
+        }
+        PyErr_Restore(ptype, pvalue, ptraceback);
+        PyErr_Clear();
+    }
+
+    if (!observed_exception) {
+        PyErr_SetString(TestError,
+                        "HeapCTypeViaMetaclassCustomNew: expected construction to fail!");
+        return NULL;
+    }
```

Comment:
This is not supposed to be here. This should be in an independent function that is called from `test_capi.py`

Reply:
ok, got it -- will fix after dinner (in ~2hr).

Reply:
@vstinner @encukou May have also some requirements regarding placement and testing approach as they have been working here more often.

Reply:
Um, well, this file is a mess. Cleaning it up (ideally, breaking it up) is on my TODO, but too low.